### ISSUE TITLE
Cannot add CURL with BUILD_SHARED_LIBS=ON

### ISSUE THREAD

**Issue by akalsi87** [ruslo/hunter#701]
I'm pasting a simple CMakeLists.txt with a LOCAL Hunter configuration:

```
# depmgr project

cmake_minimum_required(VERSION 3.0 FATAL_ERROR)

include(cmake/HunterGate.cmake)
HunterGate(
  URL "https://github.com/ruslo/hunter/archive/v0.18.2.tar.gz"
  SHA1 "3ba70e0e13bd65f80d8e984203774b0e58cbfa0f"
  LOCAL)

if (WIN32)
set(OS_NAME Windows)
elseif (APPLE)
set(OS_NAME Darwin)
else()
set(OS_NAME Unix)
endif()

if ("${CMAKE_SIZEOF_VOIDP}" EQUAL "8")
set(DEPMGR_BITNESS 64)
else()
set(DEPMGR_BITNESS 32)
endif()

if ("${BUILD_SHARED_LIBS}" EQUAL "")
  if (DEFINED ENV{BUILD_SHARED_LIBS})
    set(BUILD_SHARED_LIBS $ENV{BUILD_SHARED_LIBS})
  else()
    set(BUILD_SHARED_LIBS OFF)
  endif()
endif()

set(DEPMGR_VERSION 0.1.0)

project(depmgr VERSION ${DEPMGR_VERSION} LANGUAGES C CXX)

### get dependencies ###
hunter_add_package(ZLIB)
hunter_add_package(CURL)

find_package(ZLIB CONFIG REQUIRED)
find_package(CURL CONFIG REQUIRED)

add_library(depmgr
  arena.c
  createspec.c
  export.c
  gzip.c
  spec.c
  stream.c
  tar.c
  url.c
  zip.c
  )

add_library(DEPMGR::lib ALIAS depmgr)

target_link_libraries(depmgr PRIVATE ZLIB::zlib CURL::libcurl)
target_include_directories(depmgr
  PUBLIC .
  PRIVATE .
  )

target_compile_definitions(depmgr
  PRIVATE DM_OS=${OS_NAME}
  PRIVATE DM_BITNESS=${DEPMGR_BITNESS}
  PRIVATE DM_VER=${DEPMGR_VERSION}
  )

#add_subdirectory(main)
```

```
# cmake/Hunter/config.cmake
hunter_config(CURL VERSION 7.49.1-DEV-v4 CMAKE_ARGS BUILD_SHARED_LIBS=${BUILD_SHARED_LIBS})
hunter_config(ZLIB VERSION 1.2.8-hunter-1 CMAKE_ARGS BUILD_SHARED_LIBS=${BUILD_SHARED_LIBS})
```

I get an error:

```
-- Looking for openssl/rand.h
-- Looking for openssl/rand.h - found
-- Looking for ldap_init in ldap;dl;OpenSSL::SSL;OpenSSL::Crypto
-- Looking for ldap_init in ldap;dl;OpenSSL::SSL;OpenSSL::Crypto - not found
-- Looking for ber_init in lber;dl;OpenSSL::SSL;OpenSSL::Crypto
-- Looking for ber_init in lber;dl;OpenSSL::SSL;OpenSSL::Crypto - not found
-- Looking for include file ldap.h
-- Looking for include file ldap.h - not found
-- Looking for include file lber.h
-- Looking for include file lber.h - not found
-- LDAP_H not found CURL_DISABLE_LDAP set ON
-- LDAP needs to be enabled to support LDAPS
-- Looking for idna_to_ascii_lz in idn;dl;OpenSSL::SSL;OpenSSL::Crypto
-- Looking for idna_to_ascii_lz in idn;dl;OpenSSL::SSL;OpenSSL::Crypto - not found
-- Looking for dlopen in dl;OpenSSL::SSL;OpenSSL::Crypto
-- Looking for dlopen in dl;OpenSSL::SSL;OpenSSL::Crypto - found
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] load: /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/projects/ZLIB/hunter.cmake
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Skip '1.2.8-p3' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Skip '1.2.8-p2' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Skip '1.2.8-hunter' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] ZLIB versions available: [1.2.8-p3;1.2.8-p2;1.2.8-hunter-1;1.2.8-hunter]
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Install to: /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Install
-- [hunter] ZLIB_ROOT: /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Install (ver.: 1.2.8-hunter-1)
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Default arguments: 
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] User arguments: BUILD_SHARED_LIBS=ON
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Registering dependency:
-- [hunter *** DEBUG *** 2017-03-16T01:17:57]   package: CURL
-- [hunter *** DEBUG *** 2017-03-16T01:17:57]   depends on: ZLIB
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Making: /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL/Dependencies/ZLIB/__dep
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] Package already installed: ZLIB
-- [hunter *** DEBUG *** 2017-03-16T01:17:57] load: /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/projects/ZLIB/hunter.cmake ... end
[hunter] Custom FindZLIB module
CMake Error at /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Install/lib/cmake/ZLIB/ZLIBTargets.cmake:68 (message):
  The imported target "ZLIB::zlib" references the file

     "/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/ZLIB/Install/lib/libzd.so.1.2.8"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Install/lib/cmake/ZLIB/ZLIBTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Install/lib/cmake/ZLIB/ZLIBConfig.cmake:1 (include)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/find/FindZLIB.cmake:57 (find_package)
  CMakeLists.txt:483 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL/Build/CURL-Release-prefix/src/CURL-Release-build/CMakeFiles/CMakeOutput.log".
See also "/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL/Build/CURL-Release-prefix/src/CURL-Release-build/CMakeFiles/CMakeError.log".
CMakeFiles/CURL-Release.dir/build.make:112: recipe for target 'CURL-Release-prefix/src/CURL-Release-stamp/CURL-Release-configure' failed
make[2]: *** [CURL-Release-prefix/src/CURL-Release-stamp/CURL-Release-configure] Error 1
make[2]: Leaving directory '/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL/Build'
CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/CURL-Release.dir/all' failed
make[1]: *** [CMakeFiles/CURL-Release.dir/all] Error 2
make[1]: Leaving directory '/home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL/Build'
Makefile:86: recipe for target 'all' failed
make: *** [all] Error 2

[hunter ** FATAL ERROR **] Build step failed (dir: /home/akalsi/.hunter/_Base/3ba70e0/0f071ab/e1266bb/Build/CURL
[hunter ** FATAL ERROR **] [Directory:/home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/projects/CURL]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.external.build.failed
-------------------------------------------------------------------

CMake Error at /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/modules/hunter_fatal_error.cmake:20 (hunter_wiki)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/modules/hunter_download.cmake:544 (hunter_fatal_error)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/projects/CURL/hunter.cmake:57 (hunter_download)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.2/3ba70e0/Unpacked/cmake/modules/hunter_add_package.cmake:87 (include)
  CMakeLists.txt:39 (hunter_add_package)


-- Configuring incomplete, errors occurred!
See also "/home/akalsi/projects/depmgr/build/CMakeFiles/CMakeOutput.log".

```

**Comment by ruslo** [comment]
> HunterGate(
  URL "https://github.com/ruslo/hunter/archive/v0.18.2.tar.gz"
  SHA1 "3ba70e0e13bd65f80d8e984203774b0e58cbfa0f"
  LOCAL)

Can you please try latest version:
* https://github.com/ruslo/hunter/releases/tag/v0.18.15

**Comment by akalsi87** [comment]
This release also fails:

```
-- Looking for dlopen in dl;OpenSSL::SSL;OpenSSL::Crypto - found
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] load: /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/projects/ZLIB/hunter.cmake
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Skip '1.2.8-p3' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Skip '1.2.8-p2' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Skip '1.2.8-hunter' (looking for '1.2.8-hunter-1')
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] ZLIB versions available: [1.2.8-p3;1.2.8-p2;1.2.8-hunter-1;1.2.8-hunter]
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Install to: /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Install
-- [hunter] ZLIB_ROOT: /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Install (ver.: 1.2.8-hunter-1)
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Default arguments: 
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] User arguments: BUILD_SHARED_LIBS=ON
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Registering dependency:
-- [hunter *** DEBUG *** 2017-03-16T02:14:13]   package: CURL
-- [hunter *** DEBUG *** 2017-03-16T02:14:13]   depends on: ZLIB
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Making: /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL/Dependencies/ZLIB/__dep
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] Package already installed: ZLIB
-- [hunter *** DEBUG *** 2017-03-16T02:14:13] load: /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/projects/ZLIB/hunter.cmake ... end
[hunter] Custom FindZLIB module
CMake Error at /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Install/lib/cmake/ZLIB/ZLIBTargets.cmake:68 (message):
  The imported target "ZLIB::zlib" references the file

     "/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/ZLIB/Install/lib/libzd.so.1.2.8"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Install/lib/cmake/ZLIB/ZLIBTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Install/lib/cmake/ZLIB/ZLIBConfig.cmake:1 (include)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/find/FindZLIB.cmake:57 (find_package)
  CMakeLists.txt:483 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL/Build/CURL-Release-prefix/src/CURL-Release-build/CMakeFiles/CMakeOutput.log".
See also "/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL/Build/CURL-Release-prefix/src/CURL-Release-build/CMakeFiles/CMakeError.log".
CMakeFiles/CURL-Release.dir/build.make:112: recipe for target 'CURL-Release-prefix/src/CURL-Release-stamp/CURL-Release-configure' failed
make[2]: *** [CURL-Release-prefix/src/CURL-Release-stamp/CURL-Release-configure] Error 1
make[2]: Leaving directory '/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL/Build'
CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/CURL-Release.dir/all' failed
make[1]: *** [CMakeFiles/CURL-Release.dir/all] Error 2
make[1]: Leaving directory '/home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL/Build'
Makefile:86: recipe for target 'all' failed
make: *** [all] Error 2

[hunter ** FATAL ERROR **] Build step failed (dir: /home/akalsi/.hunter/_Base/57991b4/a17c0e8/e1266bb/Build/CURL
[hunter ** FATAL ERROR **] [Directory:/home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/projects/CURL]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.external.build.failed
-------------------------------------------------------------------

CMake Error at /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/modules/hunter_fatal_error.cmake:20 (hunter_wiki)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/modules/hunter_download.cmake:557 (hunter_fatal_error)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/projects/CURL/hunter.cmake:57 (hunter_download)
  /home/akalsi/.hunter/_Base/Download/Hunter/0.18.15/57991b4/Unpacked/cmake/modules/hunter_add_package.cmake:87 (include)
  CMakeLists.txt:39 (hunter_add_package)


-- Configuring incomplete, errors occurred!
See also "/home/akalsi/projects/depmgr/build/CMakeFiles/CMakeOutput.log".
```

**Comment by ruslo** [comment]
> This release also fails

You have to change local config too, some fixes was introduced for the packages. Try:

```cmake
hunter_config(CURL VERSION ${HUNTER_CURL_VERSION} CMAKE_ARGS BUILD_SHARED_LIBS=${BUILD_SHARED_LIBS})
hunter_config(ZLIB VERSION ${HUNTER_ZLIB_VERSION} CMAKE_ARGS BUILD_SHARED_LIBS=${BUILD_SHARED_LIBS})
```

instead.

Just for your info I got error about PIC. I think this is because OpenSSL doesn't respect BUILD_SHARED_LIBS CMake variable (somebody have to teach it).

**Comment by akalsi87** [comment]
I do see it fixed with ```${HUNTER_<PKG>_VERSION}```

Also, like you said, there is a bug in the OpenSSL ```CMakeLists.txt``` such that it respect ```${BUILD_SHARED_LIBS}```.

Are package issues filed here as well?

**Comment by ruslo** [comment]
> Are package issues filed here as well?

You mean old fixed issues about CURL/ZLIB, or about missing feature of OpenSSL?

**Comment by akalsi87** [comment]
The OpenSSL issue.It would be nice to have packages respect the ```BUILD_SHARED_LIBS``` flag as that is the accepted way to build shared libraries for clients.

**Comment by akalsi87** [comment]
Is it captured in some documentation, the current known limitations on libraries?

**Comment by ruslo** [comment]
> The OpenSSL issue

Feel free to open new issue. However from my experience if you have no plans to fix it yourself nobody will take care of it.

> the current known limitations on libraries?

OpenSSL is not CMake-based package, so you have to deal will a lot of stuff manually (iOS build, Windows build, flags from toolchain, etc.). It is a known limitation.

> Is it captured in some documentation

Feel free to add any information you find useful:
* https://github.com/ruslo/hunter/wiki/pkg.openssl

**Comment by xsacha** [comment]
This should work now after #703. Close?

**Comment by ruslo** [comment]
> This should work now after #703. Close?

Sure
